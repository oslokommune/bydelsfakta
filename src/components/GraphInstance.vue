<template>
  <div class="graph__shadow" role="tabpanel" :class="shadowClass">
    <div
      class="graph__container"
      :class="{ hidden: mode === 'table' }"
      ref="graphContainer"
      aria-hidden="true"
      @scroll="drawShadows"
      v-dragscroll.x="!isTouchDevice"
    >
      <div class="spinner" v-if="loading">
        <spinner></spinner>
        <span class="spinner__text">{{ $t('loading.loadingData') }}</span>
      </div>
      <div class="error" v-if="error">
        <h2>{{ errorMessage }}</h2>
      </div>
      <svg
        v-if="!error"
        class="graph__svg"
        aria-hidden="true"
        ref="svg"
        :class="{ loading }"
        @click="showHelp = false"
      ></svg>
    </div>
    <div
      :class="{ 'visually-hidden': mode === 'graph' }"
      class="graph__tablecontainer"
      ref="tableContainer"
      v-dragscroll.x="!isTouchDevice"
      @scroll="drawShadows"
    >
      <h3 class="table-heading"></h3>
      <table :class="{ compareDistrictsTable: compareDistricts }" ref="table">
        <thead></thead>
        <tbody></tbody>
      </table>
      <p v-if="compareDistricts" class="table-footnote">
        <small>{{ $t('graphCard.table.footnote') }}</small>
      </p>
    </div>
    <!-- <table class="visually-hidden"> -->
    <resize-observer @notify="handleResize"></resize-observer>
    <div class="help" v-if="settings.help && mode === 'graph'">
      <button
        class="help__button"
        @click="showHelp = !showHelp"
        :aria-label="showHelp ? $t('graphCard.help.ariaHide') : $t('graphCard.help.ariaShow')"
      >
        <ok-icon icon-ref="help" :options="{ size: 'small', margin: '0' }"></ok-icon>
      </button>
      <div v-if="showHelp" class="help__text">{{ settings.help }}</div>
    </div>
  </div>
</template>

<script>
import * as d3 from 'd3';
import { mapState } from 'vuex';
import { dragscroll } from 'vue-dragscroll';
import TemplateA from '../util/graph-templates/templateA';
import TemplateB from '../util/graph-templates/templateB';
import TemplateC from '../util/graph-templates/templateC';
import TemplateD from '../util/graph-templates/templateD';
import TemplateE from '../util/graph-templates/templateE';
import TemplateF from '../util/graph-templates/templateF';
import TemplateG from '../util/graph-templates/templateG';
import TemplateI from '../util/graph-templates/templateI';
import TemplateJ from '../util/graph-templates/templateJ';
import TemplateK from '../util/graph-templates/templateK';
import TemplateM from '../util/graph-templates/templateM';
import TemplateN from '../util/graph-templates/templateN';
import districtNames from '../config/districtNames';
import Spinner from '../assets/spinner.svg';
import OkIcon from './OkIcon';

export default {
  data: () => ({
    svg: false,
    shadow: {
      left: false,
      right: false,
    },
    data: null,
    loading: true,
    currentTemplate: false,
    allDistricts: false,
    error: false,
    errorMessage: '',
    showHelp: false,
  }),

  components: { OkIcon, Spinner },

  computed: {
    shadowClass() {
      let str = '';
      if (this.shadow.left && this.mode === 'graph') str += ' graph__shadow--left';
      if (this.shadow.right) str += ' graph__shadow--right';
      return str;
    },
    ...mapState(['districts', 'compareDistricts', 'isTouchDevice']),
    filteredData() {
      this.sortData(this.data);

      if (!this.compareDistricts || this.districts.includes('alle') || this.settings.noFilter) return this.data;

      const selectedDistrictNames = this.districts.map(id => districtNames[id]);

      return {
        meta: this.data.meta,
        data: this.data.data.filter(d => {
          return selectedDistrictNames.includes(d.geography) || d.totalRow;
        }),
      };
    },
  },

  watch: {
    settings: function() {
      this.draw();
    },
    districts: function(to, from) {
      if (
        from.length > 1 &&
        (to.length > 1 || this.compareDistricts) &&
        this.currentTemplate === this.settings.template
      ) {
        this.draw({ keepData: true });
      } else {
        this.draw();
      }
    },
  },

  mounted() {
    this.draw();
  },

  methods: {
    drawShadows() {
      let ref;
      if (this.mode === 'graph') {
        ref = 'graphContainer';
      } else if (this.mode === 'table') {
        ref = 'tableContainer';
      } else {
        return;
      }

      const width = this.$refs[ref].parentElement.clientWidth;
      const spaceLeft = this.$refs[ref].scrollLeft;
      const spaceRight = this.$refs[ref].scrollWidth - width - spaceLeft;

      this.shadow.left = spaceLeft > 5;
      this.shadow.right = spaceRight > 5;
    },

    handleResize() {
      if (this.loading) return;
      if (!this.svg.resize) return;
      this.svg.resize(this.data, { method: this.settings.method, compareDistricts: this.compareDistricts });
      this.drawShadows();
    },

    sortData(data) {
      const template = this.settings.template;

      data.data.sort((a, b) => {
        if (b.totalRow && a.avgRow) return -1;
        else if (a.totalRow && b.avgRow) return 1;
        else if (a.totalRow || a.avgRow) return 1;
        else if (b.totalRow || b.avgRow) return -1;
        else {
          if (template === 'a') {
            if (a.values.length && b.values.length) {
              return b.values[0][this.settings.method] - a.values[0][this.settings.method];
            }
          }

          if (template === 'f') {
            const meanA = d3.mean(a.values.flatMap(obj => [...Array(obj.value)].fill(+obj.age)));
            const meanB = d3.mean(b.values.flatMap(obj => [...Array(obj.value)].fill(+obj.age)));
            return meanA - meanB;
          }

          if (template === 'g') {
            return b.values[0] - a.values[0];
          }

          if (template === 'j' && a.values && a.values.length && b.values && b.values.length) {
            const sumA = a.values[0][this.settings.method] + a.values[1][this.settings.method];
            const sumB = b.values[0][this.settings.method] + b.values[1][this.settings.method];
            return sumB - sumA;
          }
        }
      });

      return data;
    },

    async draw(options = {}) {
      this.loading = true;

      const geoParam = this.compareDistricts ? '00' : this.districts[0];
      if (!options.keepData) {
        this.data = await d3
          .json(`${this.settings.url}?geography=${geoParam}`)
          .then(rawData => {
            const data = rawData[0];
            this.$emit('updateDate', data.meta.publishedDate);

            data.data.map(district => {
              district.noLink = !districtNames[district.id]; // add noLink flag if geography is not a district

              return district;
            });
            return data;
          })
          .catch(() => {
            this.error = true;
            this.errorMessage = this.$t('error.connectionLost');
            this.loading = false;
          });
      }

      if (this.data) {
        this.error = false;
        this.errorMessage = '';
      } else {
        return;
      }

      if (this.currentTemplate !== this.settings.template && !options.keepData) {
        switch (this.settings.template) {
          case 'a':
            this.svg = new TemplateA(this.$refs['svg']);
            break;
          case 'b':
            this.svg = new TemplateB(this.$refs['svg']);
            break;
          case 'c':
            this.svg = new TemplateC(this.$refs['svg']);
            break;
          case 'd':
            this.svg = new TemplateD(this.$refs['svg']);
            break;
          case 'e':
            this.svg = new TemplateE(this.$refs['svg']);
            break;
          case 'f':
            this.svg = new TemplateF(this.$refs['svg']);
            break;
          case 'g':
            this.svg = new TemplateG(this.$refs['svg']);
            break;
          case 'i':
            this.svg = new TemplateI(this.$refs['svg']);
            break;
          case 'j':
            this.svg = new TemplateJ(this.$refs['svg']);
            break;
          case 'k':
            this.svg = new TemplateK(this.$refs['svg']);
            break;
          case 'm':
            this.svg = new TemplateM(this.$refs['svg']);
            break;
          case 'n':
            this.svg = new TemplateN(this.$refs['svg']);
            break;
          default:
            break;
        }
      }

      this.currentTemplate = this.settings.template;

      this.loading = false;

      this.svg.render(this.filteredData, {
        method: this.settings.method,
        showPermille: this.settings.showPermille,
        mode: this.settings.mode,
        sources: this.sources,
        initialRender: true,
        compareDistricts: this.compareDistricts,
        range: '[0, 40]', // default range
        variant: this.settings.variant,
      });
      this.drawShadows();
      this.svg.setHeading(this.settings.heading);
    },
  },
  props: {
    settings: {
      type: Object,
      required: true,
    },
    sources: {
      type: Array,
      required: false,
      default: () => [],
    },
    mode: {
      type: String,
      required: true,
    },
  },
  directives: {
    dragscroll,
  },
};
</script>

<style lang="scss" scoped>
@import './../styles/colors';
@import './../styles/variables';

.graph {
  &__shadow {
    position: relative;

    &::before,
    &::after {
      position: absolute;
      top: 0;
      bottom: 0;
      z-index: 1;
      display: block;
      width: 2em;
      opacity: 0;
      transition: opacity 0.1s ease-in-out;
      content: '';
      pointer-events: none;
    }

    &::before {
      left: 0;
      background: linear-gradient(to right, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
    }

    &::after {
      right: 0;
      background: linear-gradient(to left, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
    }

    &--left::before {
      opacity: 1;
    }

    &--right::after {
      opacity: 1;
    }
  }

  &__container {
    position: relative;
    padding: 1em 1em 0 1em;
    overflow: auto;

    /deep/ &.hidden ~ .graph__dropdown {
      display: none;
    }
  }

  &__tablecontainer {
    position: relative;
    width: 100%;
    overflow-x: auto;
  }

  &__svg {
    &.loading {
      animation-name: fadeOut;
      animation-duration: 1s;
      animation-delay: 0.5s;
      animation-iteration-count: 1;
      animation-fill-mode: forwards;
      pointer-events: none;
    }
  }
}

.table-footnote {
  &.hide-footnote {
    display: none;
  }

  position: sticky;
  left: 0;
  padding: 0 1rem;

  &::before {
    content: '* ';
  }
}

.spinner {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 3;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  background-color: rgba($color-light-blue, 1);
  opacity: 0;
  animation-name: fadeIn;
  animation-duration: 1s;
  animation-delay: 0.5s;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;

  &__text {
    padding: 0.75em;
  }
}

.tick text {
  color: $color-grey-600;
  font-weight: 500;
  font-size: $font-small;
  font-family: monospace;
  text-rendering: geometricPrecision;
}

.help {
  position: absolute;
  top: 0.5em;
  right: 0.5em;
  z-index: 1;

  &__button {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 3em;
    height: 3em;
    color: $color-purple;
    border-radius: 50%;
    cursor: pointer;

    & > i {
      font-size: 1.5em;
      pointer-events: none;
    }

    &:hover {
      background: $color-blue;
    }

    &::before {
      position: absolute;
      top: -0.5rem;
      right: -0.5rem;
      z-index: -1;
      display: block;
      width: 4rem;
      height: 4rem;
      background: white;
      content: '';
    }
  }

  &__text {
    position: absolute;
    top: 3em;
    right: 0;
    width: 250px;
    padding: 1em;
    color: $color-yellow;
    font-weight: 500;
    background: $color-purple;
    border-radius: 3px;
    box-shadow: 0 3px 4px rgba($color-purple, 0.5);

    &::before {
      position: absolute;
      top: -0.5em;
      right: 0.75em;
      display: block;
      border-right: 0.75em solid transparent;
      border-bottom: 0.75em solid $color-purple;
      border-left: 0.75em solid transparent;
      content: '';
    }
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes fadeOut {
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
}
</style>

<style lang="scss">
th[data-footnote='true']::after {
  content: ' *';
}
</style>
