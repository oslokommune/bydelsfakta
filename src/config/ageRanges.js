export default [
  { range: '', label: '-- Velg --', selected: true, disabled: true },
  { range: '[0, 1]', label: '0–1 år' },
  { range: '[2, 5]', label: '2–5 år' },
  { range: '[6, 12]', label: '6–12 år' },
  { range: '[13, 15]', label: '13–15 år' },
  { range: '[16, 18]', label: '16–18 år' },
  { range: '[19, 29]', label: '19–29 år' },
  { range: '[30, 39]', label: '30–39 år' },
  { range: '[40, 49]', label: '40–49 år' },
  { range: '[50, 59]', label: '50–59 år' },
  { range: '[60, 66]', label: '60–66 år' },
  { range: '[67, 79]', label: '67–79 år' },
  { range: '[80, 89]', label: '80–89 år' },
  { range: '[90, 119]', label: '90+ år' },
];
