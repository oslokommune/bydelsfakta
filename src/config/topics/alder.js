import { apiUrl, baseUrl } from '../../util/config';
import source from './dataSources';

const API = `${apiUrl}/api/dataset`;

export default {
  text: 'Kjønn og alder',
  value: 'alder',
  production: true,
  cards: [
    {
      size: 'large',
      heading: 'Befolkningen etter alder',
      about: {
        info:
          'Statistikken viser folkemengden per 1.1. hvert år. For flere befolkningstabeller se under «Befolkning» i Oslo kommunes statistikkbank.',
        externalInfo:
          'http://statistikkbanken.oslo.kommune.no/webview/index.jsp?headers=aar&stubs=omra30&measure=common&virtualslice=antall_value&kjonn_vaslice=3&layers=kjonn_va&layers=alder_bef03&layers=virtual&study=http%3A%2F%2F192.168.101.44%3A80%2Fobj%2FfStudy%2FBe02-Befolkingen-etter-bydel-delbydel-kjonn-aldersgrupper-2017&mode=cube&v=2&virtualsubset=antall_value&omra30subset=30100%2C30101+-+30118&aarsubset=2008+-+2019&kjonn_vasubset=3&measuretype=4&alder_bef03subset=240&cube=http%3A%2F%2Fstatistikkbanken.oslo.kommune.no%3A80%2Fobj%2FfCube%2FBe02-Befolkingen-etter-bydel-delbydel-kjonn-aldersgrupper-2017_C1&alder_bef03slice=240&top=yes',
        sources: [source.ssb, source.oslo],
      },
      map: false,
      production: true,
      tabs: [
        {
          active: false,
          label: 'Antall',
          id: 'alder_segment_antall',
          help: 'Bruk slideren eller rullegardinlisten til å utforske et bestemt alderssegment.',
          template: 'ageDistribution',
          heading: 'Aldersfordeling (antall)',
          url: `${API}/alder-distribusjon-status`,
          method: 'value',
          production: true,
        },
        {
          active: false,
          label: 'Andel',
          id: 'alder_segment_andel',
          help: 'Bruk slideren eller rullegardinlisten til å utforske et bestemt alderssegment.',
          template: 'ageDistribution',
          heading: 'Aldersfordeling (andel)',
          url: `${API}/alder-distribusjon-status`,
          method: 'ratio',
          production: true,
        },
      ],
    },
    {
      size: 'large',
      heading: 'Befolkningspyramide',
      about: {
        info:
          'Statistikken viser folkemengden per 1.1. hvert år. For flere befolkningstabeller se under «Befolkning» i Oslo kommunes statistikkbank.',
        sources: [source.ssb, source.oslo],
      },
      map: false,
      production: true,
      tabs: [
        {
          active: false,
          label: 'Status',
          id: 'alder_distribusjon_status',
          template: 'pyramid',
          heading: 'Befolkningen etter alder og kjønn',
          method: 'value',
          url: `${API}/alder-distribusjon-status`,
          production: true,
        },
      ],
    },
    {
      size: 'large',
      heading: 'Gjennomsnitt- og medianalder',
      about: {
        info:
          'Statistikken viser folkemengden per 1.1. hvert år. For flere befolkningstabeller se under «Befolkning» i Oslo kommunes statistikkbank.',
        sources: [source.ssb, source.oslo],
      },
      production: true,
      map: {
        labels: ['Lavere gjennomsnittsalder', 'Høyere gjennomsnittsalder'],
        scale: [32, 43],
        reverse: true,
        method: 'avg',
        url: `${API}/alder-distribusjon-status`,
        production: true,
      },
      tabs: [
        {
          active: false,
          label: 'Status',
          id: 'alder_medianalder_status',
          method: 'total',
          heading: 'Gjennomsnitts- og medianalder',
          template: 'boxPlot',
          url: `${API}/alder-distribusjon-status`,
          production: true,
        },
      ],
    },
  ],
  options: {
    kategori: 'Befolkning',
    tema: 'Kjønn og alder',
    bgImage: `${baseUrl}/img/alder`,
    txtColor: 'rgb(245, 173, 165)',
  },
  related: ['befolkningsutvikling', 'husholdninger', 'levekaar'],
};
