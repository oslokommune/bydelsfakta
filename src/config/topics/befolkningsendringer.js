import { baseUrl } from '../../util/config';

export default {
  text: 'Fødte, døde og flytting',
  value: 'befolkningsendringer',
  production: null,
  cards: [
    {
      size: 'large',
      heading: 'Flytting etter alder',
      production: null,
      tabs: [
        {
          label: 'Status',
          method: 'value',
          heading: 'Flytting etter alder og innvandringskategori',
          template: 'm',
          url: `${baseUrl}/mockDataFlytting.json`,
          production: null,
        },
      ],
    },
  ],
  options: {
    kategori: 'Befolkning',
    tema: 'Fødte, døde og flytting',
    bgImage: `${baseUrl}/img/flytting`,
    txtColor: 'rgb(245, 173, 165)',
  },
  related: ['alder', 'befolkningsutvikling', 'innvandrerbefolkningen'],
};
