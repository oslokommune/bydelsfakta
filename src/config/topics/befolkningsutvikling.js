import { apiUrl, baseUrl } from '../../util/config';
import source from './dataSources';

const API = `${apiUrl}/api/dataset`;

export default {
  text: 'Befolkningsutvikling',
  value: 'befolkningsutvikling',
  production: true,
  cards: [
    {
      size: 'large',
      heading: 'Befolkningsutvikling',
      about:
        'Statistikken viser folkemengden per 1.1. hvert år. For flere befolkningstabeller se under «Befolkning» i Oslo kommunes statistikkbank.',
      sources: [source.ssb],
      production: true,
      tabs: [
        {
          active: false,
          label: 'Folkemengde',
          heading: 'Befolkningsutvikling',
          method: 'value',
          template: 'b',
          url: `${API}/folkemengde-utvikling-historisk`,
          production: true,
        },

        {
          active: false,
          label: 'Prosentvis endring',
          method: 'ratio',
          template: 'b',
          url: `${API}/folkemengde-utvikling-historisk-prosent`,
          production: true,
        },
      ],
    },
    {
      size: 'large',
      heading: 'Befolkningsvekst',
      about:
        'Statistikken viser folkemengden per 1.1. hvert år. For flere befolkningstabeller se under «Befolkning» i Oslo kommunes statistikkbank.',
      sources: [source.ssb],
      production: true,
      tabs: [
        {
          active: true,
          label: 'Status',
          heading: 'Befolkningsvekst',
          production: true,
          url: `${API}/nokkeltall-om-befolkningen`,
          template: 'g',
        },
      ],
    },
    {
      size: 'large',
      heading: 'Forventet utvikling',
      about:
        'Tall for historisk befolkningsutvikling er statistikk fra SSB, mens påfølgende år viser framskrevet folkemengde per 1. januar hvert år. Framskrevne tall er hentet fra Oslo kommunes seneste befolkningsframskrivning. Befolkningsframskrivningen gir et anslag for hvor mange mennesker det vil bo i Oslo i årene framover fordelt på aldersgrupper. Framskrivningene legger til grunn en rekke forutsetninger for utvikling av fruktbarhet, dødelighet, flyttemønster og boligbygging, og viser beregnet befolkningsutvikling gitt disse forutsetningene.',
      sources: [source.ssb, source.oslo],
      production: true,
      tabs: [
        {
          label: 'Historisk',
          method: 'value',
          template: 'h',
          url: `${API}/befolkningsutvikling-og-forventet-utvikling`,
          production: true,
        },
      ],
    },
  ],
  related: ['alder', 'innvandrerbefolkningen', 'befolkningsendringer'],
  options: {
    kategori: 'Befolkning',
    tema: 'Befolkningsutvikling',
    bgImage: `${baseUrl}/img/folkemengde`,
    txtColor: 'rgb(245, 173, 165)',
  },
};
