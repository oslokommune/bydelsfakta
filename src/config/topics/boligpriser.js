import { apiUrl, baseUrl } from '../../util/config';
import source from './dataSources';

const API = `${apiUrl}/api/dataset`;

export default {
  text: 'Boligpriser',
  value: 'boligpriser',
  cards: [
    {
      size: 'large',
      heading: 'Boligpris for blokkleiligheter',
      about: {
        info: '<p>Statistikken viser gjennomsnittlig kvadratmeterpris på omsatte boliger* via FINN.no i tillegg til salg registrert i grunnboken. Pris er kjøpesummen oppgitt i kjøpekontrakten. For borettslagsboliger er både innskudd og andel av fellesgjeld inkludert i prisen. Arealbegrepet som benyttes i statistikken er P-ROM. Ettersom små boliger gjerne har høyere kvadratmeterpris enn store boliger, er det kun blokkleiligheter som benyttes i dette materialet. For delbydeler med få omsatte leiligheter er dette et mindre robust mål enn der hvor det er mange omsatte leiligheter. Kvadratmeterpriser for 2018 og nyere er produsert av Virdi AS.</p> <p>*Omsatte boliger inkluderer:</p> <ol><li>Alle salg via FINN.no, og</li><li>Alle salg via grunnboken som:</li><li>står oppført som et fritt salg (i motsetning til arv e.l.)</li><li>hvor kjøper og selger ikke er samme person</li><li>salget er fullstendig (100% av eiendommen er solgt)</li><li>det er kun en enhet som er solgt (ikke en boligblokk e.l.)</li></ol><p><p>Gjennomsnittlig kvadratmeterpris:</p><li>Pris er kjøpesummen oppgitt i kjøpekontrakten. For borettslagsboliger er både innskudd og andel av fellesgjeld inkludert i prisen.</li><li>Arealbegrepet som benyttes i statistikken er P-ROM. P-ROM måles som bruksarealet innenfor boligens omsluttende vegger minus boder, tekniske rom, garasje, felles trapperom og liknende.</li><p>Boligtype:</p><li>Blokkleielighet (leiligheter i boligblokk med to etasjer eller flere)</li><li>Småhus, annet (eneboliger, rekkehus, tomannsboliger og annet)</li><p>Kvadratmeterpriser for årene 2004-2017 er produsert av SSB. Denne statistikken er basert på finn.no sitt prisregister over omsatte boliger. I tillegg supplerer SSB boligomsetningene med data fra Matrikkelen.Statistikken for 2018 og nyere er produsert av Boligmappa AS (<a href="https://www.hjemla.no">www.hjemla.no</a>). Dette medfører et brudd i tidsserien. Vi anser imidlertid sammenlignbarheten mellom de to statistikkene som temmelig god. Begge produsentene henter salgsdata fra FINN.no, men Boligmappa AS supplerer datagrunnlaget med salg fra grunnboken og har i så måte et litt større datagrunnlag enn SSB. I statistikkbanken til Oslo kommune er det tilgjengeliggjort statistikk fra både SSB og Virdi AS for årene 2018 og 2019. Gjennomsnittlig kvadratmeterpris i de to tabellene for de to overlappende årene, er ikke identiske, men avvikene er små og går i begge retninger.</p>',
        sources: [source.ssb, source.boligmappa],
        externalInfo:
          'https://statistikkbanken.oslo.kommune.no/statbank/pxweb/no/db1/db1__Boliger%20og%20byggevirksomhet__Boligpriser',
      },
      map: {
        labels: ['Lavere boligpriser', 'Høyere boligpriser'],
        heading: 'Gjennomsnittlig kvadratmeterpris blokkleiligheter',
        method: 'value',
        url: `${API}/boligpriser-blokkleiligheter-status`,
      },
      tabs: [
        {
          label: 'Status',
          id: 'boligpriser_status',
          heading: 'Gjennomsnittlig kvadratmeterpris for blokkleiligheter',
          method: 'value',
          template: 'bars',
          url: `${API}/boligpriser-blokkleiligheter-status`,
        },
        {
          label: 'Historisk',
          id: 'boligpriser_historisk',
          heading: 'Gjennomsnittlig kvadratmeterpris for blokkleiligheter',
          method: 'value',
          template: 'lines',
          url: `${API}/boligpriser-blokkleiligheter-historisk`,
        },
      ],
    },
    {
      size: 'large',
      heading: 'Median totalpris',
      about: {
        info: '<p>Statistikken viser median totalpris på omsatte boliger via FINN.no i tillegg til salg registrert i grunnboken. Pris er kjøpesummen oppgitt i kjøpekontrakten. For borettslagsboliger er både innskudd og andel av fellesgjeld inkludert i prisen.</p><p>Omsatte boliger inkluderer:</p><ol><li>Alle salg via FINN.no, og</li><li>Alle salg via grunnboken som:<ul><li>står oppført som et fritt salg (i motsetning til arv e.l.)</li><li>hvor kjøper og selger ikke er samme person</li><li>salget er fullstendig (100% av eiendommen er solgt)</li><li>det er kun en enhet som er solgt (ikke en boligblokk e.l.)</ul></ol><p>Totalpris = Salgspris + fellesgjeld<br>Median totalpris = Totalpris på boligen som er midt i rekken hvis salgene sorteres etter pris.</p>',
        sources: [source.boligmappa],
        externalInfo:
          'https://statistikkbanken.oslo.kommune.no/statbank/pxweb/no/db1/db1__Boliger%20og%20byggevirksomhet__Boligpriser',
      },
      map: {
        labels: ['Lavere boligpriser', 'Høyere boligpriser'],
        heading: 'Median totalpris',
        method: 'value',
        url: `${API}/median-totalpris-status`,
      },
      tabs: [
        {
          label: 'Status',
          id: 'median_totalpris_status',
          heading: 'Median totalpris for boliger',
          method: 'value',
          template: 'bars',
          url: `${API}/median-totalpris-status`,
        },
        {
          label: 'Historisk',
          id: 'median_totalpris_historisk',
          heading: 'Median totalpris for boliger',
          method: 'value',
          template: 'lines',
          url: `${API}/median-totalpris-historisk`,
        },
      ],
    },
  ],
  options: {
    kategori: 'Boforhold',
    tema: 'Boligpriser',
    bgImage: `${baseUrl}/img/boligpriser`,
    txtColor: 'rgb(199, 247, 201)',
  },
  related: ['rom-per-person', 'eierform', 'fodte-dode-flytting'],
};
