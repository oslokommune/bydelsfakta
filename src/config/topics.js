import alder from './topics/alder';
import befolkningsendringer from './topics/befolkningsendringer';
import boligpriser from './topics/boligpriser';
import bygningstyper from './topics/bygningstyper';
import eierform from './topics/eierform';
import befolkningsutvikling from './topics/befolkningsutvikling';
import husholdninger from './topics/husholdninger';
import innvandrerbefolkningen from './topics/innvandrerbefolkningen';
import levekaar from './topics/levekaar';
import trangboddhet from './topics/trangboddhet';

export const topicNames = [
  'befolkningsutvikling',
  'befolkningsendringer',
  'alder',
  'innvandrerbefolkningen',
  'husholdninger',
  'levekaar',
  'eierform',
  'bygningstyper',
  'boligpriser',
  'trangboddhet',
];

export const disabledTopics = ['befolkningsendringer'];

export const categories = [
  {
    kategori: 'Befolkning',
    color: 'rgb(182, 63, 50)',
    links: ['befolkningsutvikling', 'befolkningsendringer', 'alder', 'innvandrerbefolkningen', 'husholdninger'],
  },
  {
    kategori: 'Boforhold',
    color: 'rgb(27, 173, 120)',
    links: ['trangboddhet', 'eierform', 'bygningstyper', 'boligpriser'],
  },
  {
    kategori: 'Levek√•r',
    color: 'rgb(219, 160, 52)',
    links: ['levekaar'],
  },
];

export const topics = {
  alder,
  befolkningsendringer,
  boligpriser,
  bygningstyper,
  eierform,
  befolkningsutvikling,
  husholdninger,
  innvandrerbefolkningen,
  levekaar,
  trangboddhet,
};
