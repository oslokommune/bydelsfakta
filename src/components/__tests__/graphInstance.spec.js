import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import vueResize from 'vue-resize';
import GraphInstance from '../GraphInstance';
import { topics } from '../../config/topics';
import store from '../../store';
import setupI18n from '../../i18n';

const i18n = setupI18n();

describe('GraphInstance', () => {
  let wrapper = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(vueResize);
    localVue.use(Vuex);
    wrapper = mount(GraphInstance, {
      propsData: {
        settings: topics['alder'].cards[0],
        showTable: false,
        mode: 'graph',
      },
      localVue,
      store,
      i18n,
      stubs: {
        spinner: true,
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders graphInstance-component and finds graph__container-class', () => {
    expect(wrapper.classes('graph__shadow')).toBe(true);
  });

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
