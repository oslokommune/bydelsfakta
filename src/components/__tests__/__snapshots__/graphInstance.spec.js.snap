// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphInstance renders correctly 1`] = `
<div
  class="graph__shadow"
  role="tabpanel"
>
  <div
    aria-hidden="true"
    class="graph__container"
  >
    <div
      class="spinner"
    >
      <spinner-stub />
       
      <span
        class="spinner__text"
      >
        Henter data
      </span>
    </div>
     
    <!---->
     
    <svg
      aria-hidden="true"
      class="graph__svg loading"
    />
  </div>
   
  <div
    class="graph__tablecontainer visually-hidden"
  >
    <h3
      class="table-heading"
    />
     
    <table
      class=""
    >
      <thead />
       
      <tbody />
    </table>
     
    <!---->
  </div>
   
  <div
    class="resize-observer"
    data-v-2b830392=""
    tabindex="-1"
  >
    <object
      aria-hidden="true"
      data="about:blank"
      tabindex="-1"
      type="text/html"
    />
  </div>
   
  <!---->
</div>
`;
